From 38a976051c28c2a64ba2b3f1a8f2a72d32d8ed72 Mon Sep 17 00:00:00 2001
From: "Leo (Sunpeng) Li" <sunpeng.li@amd.com>
Date: Tue, 18 Jul 2017 18:33:23 -0400
Subject: [PATCH 2637/2831] drm/amd/display: Do not release state objects on
 atomic check fail

In any drm ioctl call, drm_atomic_state_clear() is called at the end to
destroy the states; even if atomic check fails. Therefore, releasing
states on atomic check failure is incorrect.

Change-Id: I26f9b1f244b171b71fd34e46df7ecf69c46c271d
Signed-off-by: Leo (Sunpeng) Li <sunpeng.li@amd.com>
Reviewed-by: Andrey Grodzovsky <Andrey.Grodzovsky@amd.com>
Acked-by: Harry Wentland <Harry.Wentland@amd.com>
---
 .../gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_types.c | 21 ++++++---------------
 1 file changed, 6 insertions(+), 15 deletions(-)

diff --git a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_types.c b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_types.c
index 600cb03..af28c27 100644
--- a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_types.c
+++ b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm_types.c
@@ -3140,7 +3140,7 @@ int amdgpu_dm_atomic_check(struct drm_device *dev,
 				conn_state = drm_atomic_get_connector_state(state, &aconnector->base);
 				if (IS_ERR(conn_state)) {
 					ret = PTR_ERR_OR_ZERO(conn_state);
-					goto fail_crtcs;
+					goto fail;
 				}
 
 				dm_conn_state = to_dm_connector_state(conn_state);
@@ -3199,7 +3199,7 @@ int amdgpu_dm_atomic_check(struct drm_device *dev,
 
 			ret = drm_atomic_add_affected_planes(state, crtc);
 			if (ret)
-				goto fail_crtcs;
+				goto fail;
 		}
 	}
 
@@ -3261,7 +3261,7 @@ int amdgpu_dm_atomic_check(struct drm_device *dev,
 					crtc_state,
 					false);
 				if (ret)
-					goto fail_planes;
+					goto fail;
 
 
 				if (dm_plane_state->surface)
@@ -3293,13 +3293,12 @@ int amdgpu_dm_atomic_check(struct drm_device *dev,
 
 		ret = do_aquire_global_lock(dev, state);
 		if (ret)
-			goto fail_planes;
-
+			goto fail;
 		WARN_ON(dm_state->context);
 		dm_state->context = dc_get_validate_context(dc, set, set_count);
 		if (!dm_state->context) {
 			ret = -EINVAL;
-			goto fail_planes;
+			goto fail;
 		}
 	}
 
@@ -3307,15 +3306,7 @@ int amdgpu_dm_atomic_check(struct drm_device *dev,
 	WARN_ON(ret);
 	return ret;
 
-fail_planes:
-	for (i = 0; i < set_count; i++)
-		for (j = 0; j < set[i].surface_count; j++)
-			dc_surface_release(set[i].surfaces[j]);
-
-fail_crtcs:
-	for (i = 0; i < set_count; i++)
-		dc_stream_release(set[i].stream);
-
+fail:
 	if (ret == -EDEADLK)
 		DRM_DEBUG_KMS("Atomic check stopped due to to deadlock.\n");
 	else if (ret == -EINTR || ret == -EAGAIN || ret == -ERESTARTSYS)
-- 
2.7.4

